### ISSUE TITLE
Bundling a library (with no “main” entry point)

### ISSUE THREAD

**Issue by mathieuleclaire** [scalacenter/scalajs-bundler#59]
The webpack key is exepecting a Main exported from JSApp. It should be nice to make it less restrictive and to allow any kind of @JSExport(ed) "main" function .

**Comment by julienrf** [comment]
I think I prefer not to have a proliferation of webpack configuration settings (and a proliferation of complex logic to handle the possible combinations of these settings), so here is a possible workaround (I didn’t try but it migth work):

~~~ scala
// build.sbt

import scalajsbundler.ScalaJSBundlerPlugin.scalaJSBundlerLauncher
import scalajsbundler.Launcher

scalaJSBundlerLauncher in fullOptJS := Launcher(fullOptJS.value.data, "")

webpackConfigFile := Some(baseDirectory.value / "webpack.config.js")
~~~

~~~ javascript
// webpack.config.js

var webpack = require("webpack");
module.exports = require("./scalajs.webpack.config")
module.exports.output.library = "YourLibrary"; // Replace “YourLibrary” with the name you want to use
~~~

The idea is to skip the launcher generated by scalajs-bundler and to directly use the output of `fullOptJS` instead. Then, in the webpack configuration file, you have to indicate that you want webpack to emit a library (since there no single entry point). You can see [here](https://webpack.github.io/docs/configuration.html#output-library) some information about how to customize the way that webpack will export your library exports.

**Comment by mathieuleclaire** [comment]
Thanks for these informations. What do you mean "Your library" ? What is it ? The name of my js function including the namespace ? What if I have several function calls depending on several server routes ?

**Comment by julienrf** [comment]
It is the name that you will use to call `JSExport`ed members.

**Comment by mathieuleclaire** [comment]
OK. And what happens if I have different calls for different routes ?

**Comment by julienrf** [comment]
You have to choose one global name that will be used by webpack to export your whole library. You will then be able to call any of the exported methods on it.

**Comment by mathieuleclaire** [comment]
It is still not clear for me. What should I choose in the following example ?

```scala
package my.package

@JSExport("MyObject")
object MyObject {

 @JSExport
  def connection(): Unit = {
...
}

 @JSExport
  def resetPassword(): Unit = {
...
  }

  @JSExport
  def run(): Unit = {
....
 }
}

**Comment by mathieuleclaire** [comment]
And ```scalaJSBundlerLauncher``` is private, so that it cannot be set in my build.sbt.
https://github.com/scalacenter/scalajs-bundler/blob/master/sbt-scalajs-bundler/src/main/scala/scalajsbundler/ScalaJSBundlerPlugin.scala#L57

**Comment by julienrf** [comment]
> It is still not clear for me. What should I choose in the following example?

Define a new name that will contain all your `JSExport`s. For instance, if you use `Foo`, webpack will generate a .js file that defines a global `Foo` object containing your exports (e.g. your `MyObject`). Thus, to call the `run` method, you will have to write the following: `Foo.MyObject().run()`.

> And `scalaJSBundlerLauncher` is private

Oops. Then you have to override `scalaJSLauncher in fullOptJS` instead. But also `webpackEntries in fullOptJS`:

~~~ scala
webpackEntries in fullOptJS := {
  val sjsOutput = fullOptJS.value.data
  Seq(sjsOutput.name.stripSuffix(".js"), sjsOutput)
}
~~~

**Comment by mathieuleclaire** [comment]
Ok for the library and ```webpakEntries```. But I don't know how to override the private TaskKey ```scalaJSBundlerLauncher```

**Comment by mathieuleclaire** [comment]
And the following code (adding a tuple in the Sequence):

```scala
webpackEntries in fullOptJS := {
      val sjsOutput = fullOptJS.value.data
      Seq((sjsOutput.name.stripSuffix(".js"), sjsOutput))
    }
```
lead to the following error:
```sh
Reference to undefined setting: 
  openmole-naked/*:fullOptJS from openmole-naked/*:fullOptJS::webpackEntries 
```

**Comment by mathieuleclaire** [comment]
I finally succeeded in. For info, here is all my config:

In built.sbt:
```scala
webpackConfigFile := Some((resourceDirectory in Compile).value / "webpack.config.js"),
webpackEntries in (Compile, fullOptJS) := {
    val sjsOutput = (fullOptJS in Compile).value.data
    Seq((sjsOutput.name.stripSuffix(".js"), sjsOutput))
  },
scalaJSLauncher in (Compile, fullOptJS) := {
    Attributed.blank[VirtualJSFile](FileVirtualJSFile((fullOptJS in Compile).value.data))
  }
```
and where webpack.config.js is

```js
var webpack = require("webpack");
module.exports = require("./scalajs.webpack.config");
module.exports.output.library = "OM";
```

And I call my js functions with the prefix ```OM.```

Well, it works. However, I am still convinced that it would be a very nice feature to render it possible transparently. I think that my use case won't be so rare !

**Comment by julienrf** [comment]
Thanks for sharing your solution. I understand that this use case might not be rare, but currently scalajs-bundler targets Scala.js projects that use NPM packages and that are executed in a browser.

I think that, to correctly support your use case we should implement something that makes it easy to publish a Scala.js project as an NPM package (that other node projects can then depend on).

**Comment by julienrf** [comment]
@mathieuleclaire I just merged #86, which documents how to achieve your goal. I think the [solution](https://github.com/scalacenter/scalajs-bundler/blob/master/sbt-scalajs-bundler/src/sbt-test/sbt-scalajs-bundler/library/build.sbt#L10-L17) is not that bad, so I would be happy to close this issue, what do you think?

**Comment by mathieuleclaire** [comment]
@julienrf it is just fine (4 years later, sorry)
